File: features/steps/user_search_steps.py
import requests
from behave import given, then, when

from config.settings import GITHUB_API_BASE
from pages.home_page import HomePage
from pages.user_details_page import UserDetailsPage


@given("I open the GitHub user search application")
def step_open_app(context):
    context.home_page = HomePage(context.page)
    context.user_details_page = UserDetailsPage(context.page)
    context.home_page.open(context.base_url)


@when('I search for github user "{username}"')
def step_search_user(context, username):
    context.searched_username = username
    context.home_page.search_user(username)


@when('I open "{username}" from the results')
def step_open_result(context, username):
    context.home_page.click_user_result(username)


@then('I should see "{username}" profile details')
def step_profile_loaded(context, username):
    context.user_details_page.wait_until_loaded(username)


@then("I should see {expected_count:d} repositories in the UI")
def step_verify_ui_repo_count(context, expected_count):
    actual_count = context.user_details_page.get_repo_count_from_ui()
    assert actual_count == expected_count, (
        f"Expected {expected_count} repositories in UI, but got {actual_count}."
    )


@then('the API should report {expected_count:d} repositories for "{username}"')
def step_verify_api_repo_count(context, expected_count, username):
    user_response = requests.get(f"{GITHUB_API_BASE}/users/{username}", timeout=10)
    user_response.raise_for_status()
    user_data = user_response.json()

    actual_count = user_data.get("public_repos")
    assert actual_count == expected_count, (
        f"Expected {expected_count} repositories from API, but got {actual_count}."
    )

File: features/steps/userSearch.steps.ts
import { expect } from '@playwright/test';
import { HomePage } from '../../pages/HomePage.js';
import { UserDetailsPage } from '../../pages/UserDetailsPage.js';
import { Given, Then, When } from './fixtures.js';

const GITHUB_API_BASE = (process.env.GITHUB_API_BASE ?? 'https://api.github.com').replace(
  /\/$/,
  '',
);

const githubRequestHeaders = {
  Accept: 'application/vnd.github+json',
  'X-GitHub-Api-Version': '2022-11-28',
  'User-Agent': 'playwright-test-bdd (BDD steps)',
  ...(process.env.GITHUB_TOKEN ? { Authorization: `Bearer ${process.env.GITHUB_TOKEN}` } : {}),
};

Given('I open the GitHub user search application', async ({ page }) => {
  const home = new HomePage(page);
  await home.open('/');
});

When('I search for github user {string}', async ({ page }, username: string) => {
  const home = new HomePage(page);
  await home.searchUser(username);
});

When('I open {string} from the results', async ({ page }, username: string) => {
  const home = new HomePage(page);
  await home.clickUserResult(username);
});

Then('I should see {string} profile details', async ({ page }, username: string) => {
  const details = new UserDetailsPage(page);
  await details.waitUntilLoaded(username);
});

Then(
  'the UI repository count should match GitHub public_repos for {string}',
  async ({ page, request }, username: string) => {
    const details = new UserDetailsPage(page);
    const apiUser = await request.get(`${GITHUB_API_BASE}/users/${username}`, {
      headers: githubRequestHeaders,
    });
    if (!apiUser.ok()) {
      throw new Error(`GitHub API ${apiUser.status()}: ${await apiUser.text()}`);
    }
    const body = (await apiUser.json()) as { public_repos?: number };
    const apiCount = body.public_repos;
    expect(apiCount, 'API public_repos missing').toBeDefined();
    const uiCount = await details.getRepoCountFromUi();
    expect(
      uiCount,
      `UI showed ${uiCount} public repos but GitHub API public_repos is ${apiCount} for '${username}'.`,
    ).toBe(apiCount);
  },
);
